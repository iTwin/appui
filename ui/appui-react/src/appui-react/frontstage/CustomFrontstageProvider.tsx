/*---------------------------------------------------------------------------------------------
 * Copyright (c) Bentley Systems, Incorporated. All rights reserved.
 * See LICENSE.md in the project root for license terms and full copyright notice.
 *--------------------------------------------------------------------------------------------*/
/** @packageDocumentation
 * @module Frontstage
 */
import type * as React from "react";
import { FrontstageProvider } from "./FrontstageProvider";
import type { StandardFrontstageProps } from "./StandardFrontstageProvider";
import { ContentGroup, ContentGroupProvider } from "../content/ContentGroup";
import type { CustomFrontstageConfig } from "./FrontstageConfig";
import type { Layout } from "../layout/Layout";

/** @internal */
export interface CustomFrontstageProviderProps
  extends Pick<StandardFrontstageProps, "id" | "contentGroupProps"> {
  content?: React.ReactNode;
  layout?: Layout;
}

/** @internal */
export class CustomFrontstageProvider extends FrontstageProvider {
  constructor(private _props: CustomFrontstageProviderProps) {
    super();
  }

  public override get id(): string {
    return this._props.id;
  }

  public override frontstageConfig(): CustomFrontstageConfig {
    const contentGroup =
      this._props.contentGroupProps instanceof ContentGroupProvider
        ? this._props.contentGroupProps
        : new ContentGroup(this._props.contentGroupProps);
    return {
      id: this._props.id,
      contentGroup,
      version: 1.0,
      content: this._props.content,
      layout: this._props.layout,
    };
  }
}
