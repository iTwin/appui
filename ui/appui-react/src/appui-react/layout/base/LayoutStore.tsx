/*---------------------------------------------------------------------------------------------
 * Copyright (c) Bentley Systems, Incorporated. All rights reserved.
 * See LICENSE.md in the project root for license terms and full copyright notice.
 *--------------------------------------------------------------------------------------------*/
/** @packageDocumentation
 * @module Base
 */

import * as React from "react";
import type { StoreApi } from "zustand";
import { createStore } from "zustand";
import { useStoreWithEqualityFn } from "zustand/traditional";
import { shallow } from "zustand/shallow";
import { assert } from "@itwin/core-bentley";
import type { NineZoneState } from "../state/NineZoneState";
import { createNineZoneState } from "../state/NineZoneState";

/** @internal */
export type LayoutState = NineZoneState;

/** @internal */
export type LayoutStore = StoreApi<NineZoneState>;

/** @internal */
export const LayoutStoreContext = React.createContext<LayoutStore | undefined>(
  undefined
);
LayoutStoreContext.displayName = "appui:LayoutStoreContext";

/** @internal */
export function createLayoutStore(args?: Partial<LayoutState>): LayoutStore {
  return createStore<LayoutState>(() => createNineZoneState(args));
}

/** @internal */
export function useLayoutStore() {
  const store = React.useContext(LayoutStoreContext);
  assert(!!store, "LayoutStoreContext is not defined");
  return store;
}

/** @internal */
export function useLayout<SelectorOutput>(
  selector: (state: LayoutState) => SelectorOutput,
  multipleSlices = false
) {
  const store = useLayoutStore();
  return useStoreWithEqualityFn(
    store,
    selector,
    multipleSlices ? shallow : undefined
  );
}
