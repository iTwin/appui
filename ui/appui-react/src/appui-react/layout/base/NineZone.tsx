/*---------------------------------------------------------------------------------------------
 * Copyright (c) Bentley Systems, Incorporated. All rights reserved.
 * See LICENSE.md in the project root for license terms and full copyright notice.
 *--------------------------------------------------------------------------------------------*/
/** @packageDocumentation
 * @module Base
 */

import * as React from "react";
import { assert, Guid } from "@itwin/core-bentley";
import {
  Rectangle,
  useRefs,
  useResizeObserver,
} from "@itwin/core-react/internal";
import type { CursorType } from "../widget-panels/CursorOverlay.js";
import type { PanelSide } from "../widget-panels/PanelTypes.js";
import { WidgetContentManager } from "../widget/ContentManager.js";
import type { FloatingWidgetResizeHandle } from "../widget/FloatingWidget.js";
import { FloatingWidget } from "../widget/FloatingWidget.js";
import {
  DraggedPanelSideContext,
  DraggedResizeHandleContext,
  DraggedWidgetIdContext,
  DragProvider,
} from "./DragManager.js";
import { WidgetTab } from "../widget/Tab.js";
import type { NineZoneAction } from "../state/NineZoneAction.js";
import type { LayoutStore } from "./LayoutStore.js";
import { LayoutStoreContext, useLayout } from "./LayoutStore.js";

/** @internal */
export type NineZoneDispatch = (action: NineZoneAction) => void;

/** @internal */
export interface NineZoneProps {
  children?: React.ReactNode;
  dispatch: NineZoneDispatch;
  layout: LayoutStore;
  labels?: NineZoneLabels;
  toolSettingsContent?: React.ReactNode;
  widgetContent?: React.ReactNode;
  tab?: React.ReactNode;
  floatingWidget?: React.ReactNode;
  showWidgetIcon?: boolean;
  autoCollapseUnpinnedPanels?: boolean;
  animateDockedToolSettings?: boolean;
  uiIsVisible?: boolean;
}

/** @internal */
export interface NineZoneLabels {
  dockToolSettingsTitle?: string;
  moreWidgetsTitle?: string;
  moreToolSettingsTitle?: string;
  pinPanelTitle?: string;
  resizeGripTitle?: string;
  sendWidgetHomeTitle?: string;
  toolSettingsHandleTitle?: string;
  unpinPanelTitle?: string;
  popoutActiveTab?: string;
}

/** @internal */
export function NineZone(props: NineZoneProps) {
  const { children, ...providerProps } = props;
  const measurerRef = React.useRef<HTMLDivElement>(null);
  const measure = React.useCallback<() => Rectangle>(() => {
    assert(!!measurerRef.current);
    return Rectangle.create(measurerRef.current.getBoundingClientRect());
  }, []);
  return (
    <NineZoneProvider measure={measure} {...providerProps}>
      <Measurer ref={measurerRef} />
      {children}
    </NineZoneProvider>
  );
}

/** @internal */
export interface NineZoneProviderProps extends NineZoneProps {
  measure: () => Rectangle;
}

const tab = <WidgetTab />;
const floatingWidget = <FloatingWidget />;

/** @internal */
export function NineZoneProvider(props: NineZoneProviderProps) {
  return (
    <LayoutStoreContext.Provider value={props.layout}>
      <NineZoneDispatchContext.Provider value={props.dispatch}>
        <NineZoneLabelsContext.Provider value={props.labels}>
          <UiIsVisibleContext.Provider value={!!props.uiIsVisible}>
            <ShowWidgetIconContext.Provider value={!!props.showWidgetIcon}>
              <AutoCollapseUnpinnedPanelsContext.Provider
                value={!!props.autoCollapseUnpinnedPanels}
              >
                <WidgetContentNodeContext.Provider value={props.widgetContent}>
                  <ToolSettingsNodeContext.Provider
                    value={props.toolSettingsContent}
                  >
                    <TabNodeContext.Provider value={props.tab || tab}>
                      <FloatingWidgetNodeContext.Provider
                        value={props.floatingWidget || floatingWidget}
                      >
                        <AnimateDockedToolSettingsContext.Provider
                          value={!!props.animateDockedToolSettings}
                        >
                          <DragProvider>
                            <CursorTypeProvider>
                              <WidgetContentManager>
                                <MeasureContext.Provider value={props.measure}>
                                  {props.children}
                                </MeasureContext.Provider>
                              </WidgetContentManager>
                            </CursorTypeProvider>
                          </DragProvider>
                        </AnimateDockedToolSettingsContext.Provider>
                      </FloatingWidgetNodeContext.Provider>
                    </TabNodeContext.Provider>
                  </ToolSettingsNodeContext.Provider>
                </WidgetContentNodeContext.Provider>
              </AutoCollapseUnpinnedPanelsContext.Provider>
            </ShowWidgetIconContext.Provider>
          </UiIsVisibleContext.Provider>
        </NineZoneLabelsContext.Provider>
      </NineZoneDispatchContext.Provider>
    </LayoutStoreContext.Provider>
  );
}

/** @internal */
export const NineZoneDispatchContext = React.createContext<NineZoneDispatch>(
  null!
);
NineZoneDispatchContext.displayName = "nz:NineZoneDispatchContext";

/** @internal */
export const NineZoneLabelsContext = React.createContext<
  NineZoneLabels | undefined
>(undefined);
NineZoneLabelsContext.displayName = "nz:NineZoneLabelsContext";

/** @internal */
export const CursorTypeContext = React.createContext<CursorType | undefined>(
  undefined
);
CursorTypeContext.displayName = "nz:CursorTypeContext";

/** @internal */
export const WidgetContentNodeContext =
  React.createContext<React.ReactNode>(undefined);
WidgetContentNodeContext.displayName = "nz:WidgetContentNodeContext";

/** @internal */
export const ShowWidgetIconContext = React.createContext<boolean>(false);
ShowWidgetIconContext.displayName = "nz:ShowWidgetIconContext";

/** @internal */
export const AutoCollapseUnpinnedPanelsContext =
  React.createContext<boolean>(false);
AutoCollapseUnpinnedPanelsContext.displayName =
  "nz:AutoCollapseUnpinnedPanelsContext";

/** @internal */
export const AnimateDockedToolSettingsContext =
  React.createContext<boolean>(false);
AnimateDockedToolSettingsContext.displayName =
  "nz:AnimateDockedToolSettingsContext";

/** @internal */
export const ToolSettingsNodeContext =
  React.createContext<React.ReactNode>(undefined);
ToolSettingsNodeContext.displayName = "nz:ToolSettingsNodeContext";

/** @internal */
export const TabNodeContext = React.createContext<React.ReactNode>(undefined);
TabNodeContext.displayName = "nz:TabNodeContext";

/** @internal */
export const FloatingWidgetNodeContext =
  React.createContext<React.ReactNode>(undefined);
FloatingWidgetNodeContext.displayName = "nz:FloatingWidgetNodeContext";

/** @internal */
export const MeasureContext = React.createContext<() => Rectangle>(null!);
MeasureContext.displayName = "nz:MeasureContext";

/** @internal */
export const UiIsVisibleContext = React.createContext<boolean>(false);
UiIsVisibleContext.displayName = "nz:UiIsVisibleContext";

function CursorTypeProvider(props: { children?: React.ReactNode }) {
  const draggedTab = useLayout((state) => !!state.draggedTab);
  const draggedWidgetId = React.useContext(DraggedWidgetIdContext);
  const draggedPanelSide = React.useContext(DraggedPanelSideContext);
  const draggedResizeHandle = React.useContext(DraggedResizeHandleContext);
  let type: CursorType | undefined;
  if (draggedTab || draggedWidgetId) type = "grabbing";
  else if (draggedPanelSide) type = sideToCursorType(draggedPanelSide);
  else if (draggedResizeHandle) type = handleToCursorType(draggedResizeHandle);
  return (
    <CursorTypeContext.Provider value={type}>
      {props.children}
    </CursorTypeContext.Provider>
  );
}

const Measurer = React.forwardRef<HTMLDivElement>(function Measurer(_, ref) {
  const size = React.useRef<{ height?: number; width?: number }>({});
  const dispatch = React.useContext(NineZoneDispatchContext);
  const handleResize = React.useCallback(
    (width: number, height: number) => {
      if (size.current.width === width && size.current.height === height)
        return;
      size.current.height = height;
      size.current.width = width;
      dispatch({
        type: "RESIZE",
        size: {
          height,
          width,
        },
      });
    },
    [dispatch]
  );
  const roRef = useResizeObserver(handleResize);
  const refs = useRefs(ref, roRef);
  return (
    <div
      ref={refs}
      style={{
        position: "absolute",
        width: "100%",
        height: "100%",
        top: "0",
        left: "0",
        visibility: "hidden",
      }}
    />
  );
});

/** @internal */
export function sideToCursorType(side: PanelSide): CursorType {
  switch (side) {
    case "bottom":
    case "top":
      return "ns-resize";
    case "left":
    case "right":
      return "ew-resize";
  }
}

/** @internal */
export function handleToCursorType(
  handle: FloatingWidgetResizeHandle
): CursorType {
  switch (handle) {
    case "bottom":
    case "top":
      return "ns-resize";
    case "left":
    case "right":
      return "ew-resize";
    case "topLeft":
    case "bottomRight":
      return "nwse-resize";
    case "topRight":
    case "bottomLeft":
      return "nesw-resize";
  }
}

/** @internal */
export function getUniqueId() {
  return Guid.createValue();
}

/** @internal */
export function useLabel(labelKey: keyof NineZoneLabels) {
  const labels = React.useContext(NineZoneLabelsContext);
  return labels?.[labelKey];
}
