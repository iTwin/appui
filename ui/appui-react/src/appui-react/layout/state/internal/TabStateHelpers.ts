/*---------------------------------------------------------------------------------------------
 * Copyright (c) Bentley Systems, Incorporated. All rights reserved.
 * See LICENSE.md in the project root for license terms and full copyright notice.
 *--------------------------------------------------------------------------------------------*/
/** @packageDocumentation
 * @module Base
 */

import type { Draft } from "immer";
import { castDraft, produce } from "immer";
import { UiError } from "@itwin/appui-abstract";
import type { NineZoneState } from "../NineZoneState.js";
import type { DraggedTabState, TabState } from "../TabState.js";
import { category } from "./NineZoneStateHelpers.js";
import type { SavedTabState, TabHomeState } from "../SavedTabState.js";
import { getTabLocation } from "../TabLocation.js";
import {
  addFloatingWidget,
  assertWidgetState,
  getWidgetState,
  removeWidget,
  setWidgetActiveTabId,
  updateWidgetState,
} from "./WidgetStateHelpers.js";
import type { WidgetState } from "../WidgetState.js";
import { isFloatingWidgetRestoreState } from "../WidgetRestoreState.js";
import { Rectangle } from "@itwin/core-react/internal";
import { getUniqueId } from "../../base/NineZone.js";
import { insertPanelWidget } from "./PanelStateHelpers.js";

/** @internal */
export function createTabState(
  id: TabState["id"],
  args?: Partial<TabState>
): TabState {
  return {
    label: "",
    ...args,
    id,
    unloaded: false,
  };
}

/** @internal */
export function updateTabState(
  state: NineZoneState,
  id: TabState["id"],
  update: (draft: Draft<TabState>) => void
) {
  // eslint-disable-next-line deprecation/deprecation
  if (!(id in state.tabs)) throw new UiError(category, "Tab does not exist");

  return produce(state, (draft) => {
    const tab = draft.tabs[id];
    update(tab);
  });
}

/** @internal */
export function updateSavedTabState(
  state: NineZoneState,
  id: TabState["id"],
  update: (draft: Draft<SavedTabState>) => void
) {
  return produce(state, (draft) => {
    const allIds = draft.savedTabs.allIds;
    const byId = draft.savedTabs.byId;
    let tab = byId[id];
    if (!tab) {
      allIds.push(id);
      tab = byId[id] = { id };
    } else {
      const index = allIds.indexOf(id);
      allIds.splice(index, 1);
      allIds.push(id);
    }

    update(tab);
  });
}

/** Adds a new `tab`.
 * @internal
 */
export function addTab(
  state: NineZoneState,
  id: TabState["id"],
  tabArgs?: Partial<TabState>
): NineZoneState {
  // eslint-disable-next-line deprecation/deprecation
  if (id in state.tabs) throw new UiError(category, "Tab already exists");
  const tab = {
    ...createTabState(id),
    ...tabArgs,
  };
  return produce(state, (stateDraft) => {
    stateDraft.tabs[id] = castDraft(tab);
  });
}

/** Adds an existing `tab` to a specified `widget`.
 * @internal
 */
export function addTabToWidget(
  state: NineZoneState,
  tabId: TabState["id"],
  widgetId: WidgetState["id"]
): NineZoneState {
  return insertTabToWidget(state, tabId, widgetId, Infinity);
}

/** Inserts an existing `tab` to a specified `widget` at a specified `tabIndex`.
 * @internal
 */
export function insertTabToWidget(
  state: NineZoneState,
  tabId: TabState["id"],
  widgetId: WidgetState["id"],
  tabIndex: number
): NineZoneState {
  if (!(tabId in state.tabs))
    // eslint-disable-next-line deprecation/deprecation
    throw new UiError(category, "Tab does not exist", undefined, () => ({
      tabId,
    }));
  assertWidgetState(state, widgetId);
  const location = getTabLocation(state, tabId);
  if (location)
    // eslint-disable-next-line deprecation/deprecation
    throw new UiError(
      category,
      "Tab is already in a widget",
      undefined,
      () => ({ tabId, widgetId: location.widgetId })
    );

  return produce(state, (draft) => {
    const widget = getWidgetState(draft, widgetId);
    widget.tabs.splice(tabIndex, 0, tabId);
  });
}

/** Removes tab from the UI, but keeps the tab state.
 * @internal
 */
export function removeTabFromWidget(
  state: NineZoneState,
  tabId: TabState["id"]
): NineZoneState {
  const location = getTabLocation(state, tabId);
  if (!location) return state;

  const widgetId = location.widgetId;
  const widget = getWidgetState(state, widgetId);
  const tabs = [...widget.tabs];
  const tabIndex = tabs.indexOf(tabId);
  tabs.splice(tabIndex, 1);

  if (tabs.length === 0) {
    return removeWidget(state, widgetId);
  }

  if (tabId === widget.activeTabId) {
    state = setWidgetActiveTabId(state, widget.id, tabs[0]);
  }

  return updateWidgetState(state, widgetId, {
    tabs,
  });
}

/** Removes tab from the UI and deletes the tab state.
 * @internal
 */
export function removeTab(
  state: NineZoneState,
  tabId: TabState["id"]
): NineZoneState {
  // eslint-disable-next-line deprecation/deprecation
  if (!(tabId in state.tabs)) throw new UiError(category, "Tab does not exist");

  state = removeTabFromWidget(state, tabId);
  return produce(state, (draft) => {
    delete draft.tabs[tabId];

    if (draft.toolSettings?.tabId === tabId) {
      draft.toolSettings = undefined;
    }
  });
}

const defaultHomeState: TabHomeState = {
  side: "left",
  widgetId: "",
  widgetIndex: 0,
  tabIndex: 0,
};

/** Adds removed tab to the UI.
 * @internal
 */
export function addRemovedTab(
  state: NineZoneState,
  tabId: TabState["id"]
): NineZoneState {
  if (!(tabId in state.tabs))
    // eslint-disable-next-line deprecation/deprecation
    throw new UiError(category, "Tab does not exist", undefined, () => ({
      tabId,
    }));

  const savedTab = state.savedTabs.byId[tabId];
  const home = savedTab?.home || defaultHomeState;
  const { tabIndex, widgetId } = home;

  // Add to an existing widget (by widget id).
  if (widgetId in state.widgets) {
    return insertTabToWidget(state, tabId, widgetId, tabIndex);
  }

  // Add to a floating widget.
  if (isFloatingWidgetRestoreState(home)) {
    // Add to a new floating widget.
    const nzBounds = Rectangle.createFromSize(state.size);
    const bounds = Rectangle.create(home.floatingWidget.bounds).containIn(
      nzBounds
    );
    return addFloatingWidget(state, widgetId, [tabId], {
      ...home.floatingWidget,
      bounds: bounds.toProps(),
    });
  }

  // Add to a panel section.
  const panel = state.panels[home.side];

  // Add to existing panel section.
  if (panel.widgets.length >= panel.maxWidgetCount) {
    const sectionIndex = Math.min(panel.maxWidgetCount - 1, home.widgetIndex);
    const sectionId = panel.widgets[sectionIndex];
    return insertTabToWidget(state, tabId, sectionId, home.tabIndex);
  }

  // Create a new panel section.
  const newSectionId = getUniqueId();
  return insertPanelWidget(
    state,
    panel.side,
    newSectionId,
    [tabId],
    home.widgetIndex
  );
}
