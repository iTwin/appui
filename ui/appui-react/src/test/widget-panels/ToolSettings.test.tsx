/*---------------------------------------------------------------------------------------------
 * Copyright (c) Bentley Systems, Incorporated. All rights reserved.
 * See LICENSE.md in the project root for license terms and full copyright notice.
 *--------------------------------------------------------------------------------------------*/
import { Rectangle } from "@itwin/core-react";
import { render, screen } from "@testing-library/react";
import { act, renderHook } from "@testing-library/react-hooks";
import * as React from "react";
import type { ToolSettingsEntry } from "../../appui-react";
import {
  ConfigurableCreateInfo,
  FrontstageDef,
  ToolSettingsContent,
  ToolSettingsDockedContent,
  ToolSettingsGrid,
  ToolUiProvider,
  UiFramework,
  useHorizontalToolSettingEntries,
  useToolSettingsNode,
  WidgetDef,
  WidgetPanelsToolSettings,
} from "../../appui-react";
import { InternalFrontstageManager } from "../../appui-react/frontstage/InternalFrontstageManager";
import {
  DragManager,
  DragManagerContext,
} from "../../appui-react/layout/base/DragManager";
import { createLayoutStore } from "../../appui-react/layout/base/LayoutStore";
import { NineZoneProvider } from "../../appui-react/layout/base/NineZone";
import { createNineZoneState } from "../../appui-react/layout/state/NineZoneState";
import { addTab } from "../../appui-react/layout/state/internal/TabStateHelpers";
import {
  addDockedToolSettings,
  addWidgetToolSettings,
} from "../../appui-react/layout/state/internal/ToolSettingsStateHelpers";
import { addFloatingWidget } from "../../appui-react/layout/state/internal/WidgetStateHelpers";
import { childStructure } from "../TestUtils";

describe("WidgetPanelsToolSettings", () => {
  it("should not render w/o tool settings top center zone", () => {
    vi.spyOn(
      UiFramework.frontstages,
      "activeFrontstageDef",
      "get"
    ).mockImplementation(() => undefined);
    const { container } = render(
      <NineZoneProvider
        dispatch={vi.fn()}
        layout={createLayoutStore()}
        measure={vi.fn()}
      >
        <WidgetPanelsToolSettings />
      </NineZoneProvider>
    );
    expect(container.innerHTML).toEqual("");
  });

  it("should render", () => {
    const frontstageDef = new FrontstageDef();
    const toolSettings = new WidgetDef();
    vi.spyOn(
      UiFramework.frontstages,
      "activeFrontstageDef",
      "get"
    ).mockImplementation(() => frontstageDef);
    vi.spyOn(
      InternalFrontstageManager,
      "activeToolSettingsProvider",
      "get"
    ).mockImplementation(() => undefined);
    vi.spyOn(frontstageDef, "toolSettings", "get").mockImplementation(
      () => toolSettings
    );
    let state = createNineZoneState();
    state = addTab(state, "ts");
    state = addDockedToolSettings(state, "ts");
    const sut = render(
      <NineZoneProvider
        dispatch={vi.fn()}
        layout={createLayoutStore(state)}
        measure={vi.fn()}
      >
        <WidgetPanelsToolSettings />
      </NineZoneProvider>
    );
    sut.getByText(/tools.noToolSettingsStart/);
  });
});

describe("ToolSettingsDockedContent", () => {
  class ToolUiProviderMock extends ToolUiProvider {
    constructor(info: ConfigurableCreateInfo, options: any) {
      super(info, options);
    }
  }

  it("should render settings", () => {
    const activeToolSettingsProvider = new ToolUiProviderMock(
      new ConfigurableCreateInfo("test", "test", "test"),
      undefined
    );
    vi.spyOn(
      InternalFrontstageManager,
      "activeToolSettingsProvider",
      "get"
    ).mockImplementation(() => activeToolSettingsProvider);
    const horizontalToolSettingNodes: ToolSettingsEntry[] = [
      { labelNode: "Date", editorNode: <input type="date" /> },
    ];
    vi.spyOn(
      activeToolSettingsProvider,
      "horizontalToolSettingNodes",
      "get"
    ).mockImplementation(() => horizontalToolSettingNodes);
    const sut = render(
      <DragManagerContext.Provider value={new DragManager()}>
        <ToolSettingsDockedContent />
      </DragManagerContext.Provider>
    );
    sut.getByText("Date");
  });
});

describe("ToolSettingsGrid", () => {
  it("should render", () => {
    const entries: ToolSettingsEntry[] = [
      { labelNode: "Date", editorNode: <input type="date" /> },
    ];

    render(<ToolSettingsGrid settings={entries} />);
    expect(
      screen.getByText("Date", {
        selector:
          ".uifw-standard-toolsettings-two-column-grid .uifw-standard-toolsettings-label-entry",
      }).parentElement
    ).to.satisfy(childStructure("input[type='date']"));
  });
});

describe("ToolSettingsContent", () => {
  class ToolUiProviderMock extends ToolUiProvider {
    constructor(info: ConfigurableCreateInfo, options: any) {
      super(info, options);
    }
  }

  it("should not render if not in 'widget' mode", () => {
    const { container } = render(
      <NineZoneProvider
        dispatch={vi.fn()}
        measure={vi.fn()}
        layout={createLayoutStore()}
      >
        <ToolSettingsContent />
      </NineZoneProvider>
    );
    expect(container.firstChild).to.be.null;
  });

  it("should render (Floating Widget mode)", () => {
    const activeToolSettingsProvider = new ToolUiProviderMock(
      new ConfigurableCreateInfo("test", "test", "test"),
      undefined
    );
    vi.spyOn(
      InternalFrontstageManager,
      "activeToolSettingsProvider",
      "get"
    ).mockImplementation(() => activeToolSettingsProvider);
    vi.spyOn(
      activeToolSettingsProvider,
      "toolSettingsNode",
      "get"
    ).mockImplementation(() => <div>Hello World</div>);
    let state = createNineZoneState();
    state = addTab(state, "ts");
    state = addFloatingWidget(state, "fw1", ["ts"]);
    state = addWidgetToolSettings(state, "ts");
    const layout = createLayoutStore(state);
    const component = render(
      <NineZoneProvider
        layout={layout}
        dispatch={vi.fn()}
        measure={() => new Rectangle()}
      >
        <div className="nz-floating-toolsettings">
          <ToolSettingsContent />
        </div>
      </NineZoneProvider>
    );
    component.getByText("Hello World");
  });
});

describe("useHorizontalToolSettingEntries", () => {
  it("should add tool activated event listener", () => {
    const addListenerSpy = vi.spyOn(
      UiFramework.frontstages.onToolActivatedEvent,
      "addListener"
    );
    const removeListenerSpy = vi.spyOn(
      UiFramework.frontstages.onToolActivatedEvent,
      "removeListener"
    );
    const sut = renderHook(() => useHorizontalToolSettingEntries());
    sut.unmount();
    expect(addListenerSpy).toHaveBeenCalledOnce();
    expect(removeListenerSpy).toHaveBeenCalledOnce();
  });

  it("should add tool settings reload event listener", () => {
    const addListenerSpy = vi.spyOn(
      UiFramework.frontstages.onToolSettingsReloadEvent,
      "addListener"
    );
    const removeListenerSpy = vi.spyOn(
      UiFramework.frontstages.onToolSettingsReloadEvent,
      "removeListener"
    );
    const sut = renderHook(() => useHorizontalToolSettingEntries());
    act(() => {
      UiFramework.frontstages.onToolSettingsReloadEvent.emit();
    });
    sut.unmount();
    expect(addListenerSpy).toHaveBeenCalledOnce();
    expect(removeListenerSpy).toHaveBeenCalledOnce();
  });

  it("should return undefined if activeToolSettingsProvider is unset", () => {
    const { result } = renderHook(() => useHorizontalToolSettingEntries());
    vi.spyOn(
      InternalFrontstageManager,
      "activeToolSettingsProvider",
      "get"
    ).mockReturnValue(undefined);
    act(() => {
      UiFramework.frontstages.onToolActivatedEvent.emit({ toolId: "t1" });
    });
    expect(result.current).toEqual(undefined);
  });

  it("should update tool settings", () => {
    const node = <></>;
    const entries: ToolSettingsEntry[] = [
      { labelNode: "Date", editorNode: <input type="date" /> },
    ];

    class Tool1UiProvider extends ToolUiProvider {
      constructor(info: ConfigurableCreateInfo, options: any) {
        super(info, options);
        this.toolSettingsNode = node;
        this.horizontalToolSettingNodes = this.getHorizontalToolSettings();
      }

      private getHorizontalToolSettings(): ToolSettingsEntry[] | undefined {
        return entries;
      }
    }

    vi.spyOn(
      InternalFrontstageManager,
      "activeToolSettingsProvider",
      "get"
    ).mockImplementation(
      () =>
        new Tool1UiProvider(
          new ConfigurableCreateInfo("test", "test", "test"),
          undefined
        )
    );
    const sut = renderHook(() => useHorizontalToolSettingEntries());

    act(() => {
      vi.spyOn(
        InternalFrontstageManager,
        "activeToolSettingsProvider",
        "get"
      ).mockImplementation(
        () =>
          new Tool1UiProvider(
            new ConfigurableCreateInfo("test", "test", "test"),
            undefined
          )
      );
      UiFramework.frontstages.onToolActivatedEvent.emit({
        toolId: "",
      });
      UiFramework.frontstages.onToolSettingsReloadEvent.emit();
    });

    expect(sut.result.current).toEqual(entries);
  });
});

describe("useToolSettingsNode", () => {
  class ToolUiProviderMock extends ToolUiProvider {
    constructor(info: ConfigurableCreateInfo, options: any) {
      super(info, options);
    }
  }

  it("should add/remove tool activated event listener", () => {
    const addListenerSpy = vi.spyOn(
      UiFramework.frontstages.onToolActivatedEvent,
      "addListener"
    );
    const removeListenerSpy = vi.spyOn(
      UiFramework.frontstages.onToolActivatedEvent,
      "removeListener"
    );
    const sut = renderHook(() => useToolSettingsNode());
    sut.unmount();
    expect(addListenerSpy).toHaveBeenCalledOnce();
    expect(removeListenerSpy).toHaveBeenCalledOnce();
  });

  it("should add/remove tool settings reload event listener", () => {
    const addListenerSpy = vi.spyOn(
      UiFramework.frontstages.onToolSettingsReloadEvent,
      "addListener"
    );
    const removeListenerSpy = vi.spyOn(
      UiFramework.frontstages.onToolSettingsReloadEvent,
      "removeListener"
    );
    const sut = renderHook(() => useToolSettingsNode());
    UiFramework.frontstages.onToolSettingsReloadEvent.emit();
    sut.unmount();
    expect(addListenerSpy).toHaveBeenCalledOnce();
    expect(removeListenerSpy).toHaveBeenCalledOnce();
  });

  it("should update toolSettingsNode", () => {
    const activeToolSettingsProvider = new ToolUiProviderMock(
      new ConfigurableCreateInfo("test", "test", "test"),
      undefined
    );
    vi.spyOn(
      InternalFrontstageManager,
      "activeToolSettingsProvider",
      "get"
    ).mockImplementation(() => activeToolSettingsProvider);
    const sut = renderHook(() => useToolSettingsNode());

    const node = <div>Hello World</div>;
    act(() => {
      vi.spyOn(
        activeToolSettingsProvider,
        "toolSettingsNode",
        "get"
      ).mockImplementation(() => node);
      UiFramework.frontstages.onToolActivatedEvent.emit({
        toolId: "",
      });
      UiFramework.frontstages.onToolSettingsReloadEvent.emit();
    });

    expect(sut.result.current).toEqual(node);
  });

  it("should initialize to undefined w/o active activeToolSettingsProvider", () => {
    vi.spyOn(
      InternalFrontstageManager,
      "activeToolSettingsProvider",
      "get"
    ).mockImplementation(() => undefined);
    const { result } = renderHook(() => useToolSettingsNode());

    expect(result.current).toEqual(undefined);
  });

  it("should return undefined if activeToolSettingsProvider is unset", () => {
    const { result } = renderHook(() => useToolSettingsNode());
    act(() => {
      UiFramework.frontstages.setActiveToolId("t1");
    });
    expect(result.current).toEqual(undefined);
  });
});
