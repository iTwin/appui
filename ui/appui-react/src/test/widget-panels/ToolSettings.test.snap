// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ToolSettingsContent should render (Floating Widget mode) 1`] = `
<div
  class="nz-floating-toolsettings"
  style="visibility: visible;"
>
  <div
    class="uifw-floating-toolsettings-container"
    data-toolsettings-provider="ToolUiProviderMock[test]"
  >
    <div
      class="nz-widget-content"
      data-item-type="widget-content"
    >
      <div>
        Hello World
      </div>
    </div>
  </div>
</div>
`;

exports[`ToolSettingsDockedContent should render settings 1`] = `
<div
  class="nz-toolSettings-docked"
  data-toolsettings-provider="ToolUiProviderMock[test]"
>
  <div
    class="nz-toolSettings-handle"
  >
    <svg
      fill="var(--iui-color-icon-muted, currentColor)"
      height="1rem"
      viewBox="0 0 16 16"
      width="1rem"
    >
      <path
        d="M7 4a1 1 0 11-2 0 1 1 0 012 0zM6 7a1 1 0 100 2 1 1 0 000-2zm0 4a1 1 0 100 2 1 1 0 000-2zm5-7a1 1 0 11-2 0 1 1 0 012 0zm-1 3a1 1 0 100 2 1 1 0 000-2zm0 4a1 1 0 100 2 1 1 0 000-2z"
      />
    </svg>
  </div>
  <div>
    <div
      class="nz-toolSettings-setting"
    >
      <div
        class="uif-toolsetting-label-docked-horizontal"
      >
        Date
      </div>
      <input
        type="date"
      />
    </div>
  </div>
</div>
`;

exports[`WidgetPanelsToolSettings should render 1`] = `
<div
  class="nz-toolSettings-docked"
  data-toolsettings-provider="none"
>
  <div
    class="nz-toolSettings-handle"
  >
    <svg
      fill="var(--iui-color-icon-muted, currentColor)"
      height="1rem"
      viewBox="0 0 16 16"
      width="1rem"
    >
      <path
        d="M7 4a1 1 0 11-2 0 1 1 0 012 0zM6 7a1 1 0 100 2 1 1 0 000-2zm0 4a1 1 0 100 2 1 1 0 000-2zm5-7a1 1 0 11-2 0 1 1 0 012 0zm-1 3a1 1 0 100 2 1 1 0 000-2zm0 4a1 1 0 100 2 1 1 0 000-2z"
      />
    </svg>
  </div>
  <div>
    <div
      class="nz-toolSettings-setting"
    >
      <div
        class="uif-toolsetting-label-docked-horizontal"
      >
        <div
          class="uif-toolsetting-label-docked-horizontal-empty"
        >
          tools.noToolSettingsStart
          tools.noToolSettingsPlaceholderName
          tools.noToolSettingsEnd
        </div>
      </div>
      <div />
    </div>
  </div>
</div>
`;
